stl/inc/flat_map

```diff
@@ -331,36 +337,24 @@ public:
         return *this;
     }
 
-    _NODISCARD iterator begin() noexcept {
-        return iterator {_Data.keys.cbegin(), _Data.values.begin()};
-    }
-
-    _NODISCARD const_iterator begin() const noexcept {
-        return const_iterator {_Data.keys.cbegin(), _Data.values.cbegin()};
-    }
-
-    _NODISCARD iterator end() noexcept {
-        return iterator {_Data.keys.cend(), _Data.values.end()};
-    }
-
-    _NODISCARD const_iterator end() const noexcept {
-        return const_iterator {_Data.keys.cend(), _Data.values.cend()};
-    }
-
-    _NODISCARD reverse_iterator rbegin() noexcept {
-        return _STD make_reverse_iterator(end());
+    template <class _Self_t>
+    _NODISCARD auto begin(this _Self_t&& _Self) noexcept {
+        return _Maybe_const_iterator_t<_Self_t&&> { _Self._Data.keys.cbegin(), _Self._Data.values.begin() };

```

Comment:
This refactor doesn't seem correct to me, as it introduce unexpected signatures like `begin()&&` and `begin()const&&`

Reply:
@frederick-vs-ja
I'm not confident about this, is implementation allowed to offer extra `& vs &&` overload for standard methods like this?

Reply:
This is allowed by [[member.functions]/2](https://eel.is/c++draft/member.functions#2). Without the allowance `std::source_location::current` can't be implemented.

Reply:
Refactored the code. It no longer uses deducing this.